piker/data/_web_bs.py

```diff
@@ -219,20 +225,49 @@ async def recv_task():
             '''
             receives every ws message and stores it in its corresponding result
             field, then sets the event to wakeup original sender tasks.
+            also recieves responses to requests originated from the server side.
+            reconnects the tasks after timeout.
             '''
-            async for msg in ws:
-                msg = dtype(**msg)
-
-                if msg.id not in rpc_results:
-                    log.warning(f'Wasn\'t expecting ws msg: {json.dumps(msg, indent=4)}')
-
-                res = rpc_results.setdefault(
-                    msg.id,
-                    {'result': None, 'event': trio.Event()}
-                )
-
-                res['result'] = msg
-                res['event'].set()
+            with trio.move_on_after(timeout) as cancel_scope:
+                async for msg in ws:
+                    match msg:
+                        case {
+                            'result': result,
+                            'id': mid,
+                        } if res_entry := rpc_results.get(mid):
+
+                            res_entry['result'] = response_type(**msg)
+                            res_entry['event'].set()
+
+                        case {
+                            'result': _,
+                            'id': mid,
+                        } if not rpc_results.get(mid):
+                            log.warning(f'Wasn\'t expecting ws msg: {json.dumps(msg, indent=4)}')
+
+                        case {
+                            'method': _,
+                            'params': _,
+                        }:
+                            log.debug(f'Recieved\n{msg}')
+                            if request_hook:
+                                await request_hook(request_type(**msg))
+
+                        case {
+                            'error': error
+                        }:
+                            log.warning(f'Recieved\n{error}')
+                            if error_hook:
+                                await error_hook(response_type(**msg))
+
+                        case _:
+                            log.warning(f'Unhandled JSON-RPC msg!?\n{msg}')
+
+            if cancel_scope.cancelled_caught:
+                await ws._connect()
```

Comment:
I think this whole recursively spawning tasks on timeout is a little confusing.

questions:
- if **this** task has timed out it shouldn't matter if the next try is scheduled async right?
- on timeout you run your `timeout_hook()` handler, does that need to be run async? if not, why not just do this all sync in an inf loop -> use a `while ws.connected()` (or wtv) and do a `break` once you get a result?

Further I think you could avoid passing a `timeout_args` and instead just make `timeout_hook` a `functools.partial` in the case you need to bind inputs.